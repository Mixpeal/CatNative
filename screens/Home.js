import React, { useState, useEffect, useCallback } from 'react';
import axios from 'axios';
import { View, Text, FlatList, RefreshControl, ActivityIndicator, TouchableWithoutFeedback } from 'react-native'
import EachCat from '../components/cats/EachCat';
import { useIsFocused } from '@react-navigation/native';
import Toast from 'react-native-simple-toast';


export default function Home() {
    const [refreshing, setRefreshing] = useState(false);
    const [loading, setLoading] = useState(false);
    const [loaded, setLoaded] = useState(false);
    const [hasError, setHasError] = useState(false);
    const [cats, setCats] = useState([]);
    const [page, setPage] = useState(0);
    const isFocused = useIsFocused();

    const handleResult = (result) => {
        setLoading(false);
        setPage(page + 1)
        setCats([...cats, ...result.data]);
        if (result.data && result.data.length == 0) {
            setLoaded(true)
            Toast.show('End of the line ðŸŽ‰');
        }
    }
    const fetchData = async () => {
        if (!loading && !loaded) {
            setHasError(false)
            setLoading(true);
            await axios.get(`https://api.thecatapi.com/v1/breeds?limit=20&page=${page}`)
                .then(result => handleResult(result))
                .catch(error => setHasError(true))
        }
    }
    const onRefresh = useCallback(() => {
        setRefreshing(true);
        setPage(0)
        setLoaded(false)
        fetchData().then(() => setRefreshing(false));
    }, [isFocused]);


    useEffect(() => {
        fetchData();
    }, [isFocused])

    return (
        <View style={{ backgroundColor: '#fff', flex: 1 }}>
            {!hasError && cats.length > 0 ?
                <FlatList
                    initialNumToRender={20}
                    data={cats}
                    testID="cat-list"
                    renderItem={({ item }) => <EachCat item={item} />}
                    keyExtractor={(item) => "item_" + item['id']}
                    onEndReachedThreshold={0.01}
                    onEndReached={(info) => fetchData(info)}
                    refreshControl={
                        <RefreshControl
                            refreshing={refreshing}
                            onRefresh={onRefresh}
                        />
                    }
                /> :

                <View style={{ flex: 1, flexDirection: 'column', justifyContent: 'center', alignItems: 'center' }}>
                    {(loading || refreshing) && !hasError ? (
                        <View style={{ paddingVertical: 10 }}>
                            <ActivityIndicator testId="loading-indicator" size={40} color="red" />
                        </View>
                    ) : <View>
                        <Text style={{ color: '#d3d3d4', textAlign: 'center', marginBottom: 20 }}>Something went wrong</Text>
                        <TouchableWithoutFeedback onPress={onRefresh}>
                            <Text style={{ color: '#000', fontWeight: 'bold', textAlign: 'center' }} testID="tapToRetry" >Tap to retry</Text>
                        </TouchableWithoutFeedback>
                    </View>}

                </View>
            }

            {(loading || refreshing) && !hasError && cats.length > 0 ? (
                <View style={{ paddingVertical: 10 }}>
                    <ActivityIndicator color="red" />
                </View>
            ) : null}
        </View >
    );
}

