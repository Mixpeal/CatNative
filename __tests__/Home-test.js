import 'react-native';
import React from 'react'
import Home from "../screens/Home";
import renderer from 'react-test-renderer';

it('renders correctly', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
});