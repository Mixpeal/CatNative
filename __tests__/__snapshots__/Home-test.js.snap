// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#d3d3d4",
            "marginBottom": 20,
            "textAlign": "center",
          }
        }
      >
        Something went wrong
      </Text>
      <Text
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "color": "#000",
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
        testID="tapToRetry"
      >
        Tap to retry
      </Text>
    </View>
  </View>
</View>
`;
